File: kompics/kompicsjava/src/main/java/se/kth/benchmarks/kompicsjava/bench/chameneos/ChameneosColour.java
package se.kth.benchmarks.kompicsjava.bench.chameneos;

public enum ChameneosColour {
    RED, YELLOW, BLUE, FADED;

    public static ChameneosColour forId(int i) {
        switch (i % 3) {
        case 0:
            return RED;
        case 1:
            return YELLOW;
        case 2:
            return BLUE;
        default:
            throw new RuntimeException("Unreachable!");
        }
    }

    public ChameneosColour complement(ChameneosColour other) {
        switch (this) {
        case RED:
            switch (other) {
            case RED:
                return RED;
            case YELLOW:
                return BLUE;
            case BLUE:
                return YELLOW;
            case FADED:
                return FADED;
            }
            break;
        case YELLOW:
            switch (other) {
            case RED:
                return BLUE;
            case YELLOW:
                return YELLOW;
            case BLUE:
                return RED;
            case FADED:
                return FADED;
            }
            break;
        case BLUE:
            switch (other) {
            case RED:
                return YELLOW;
            case YELLOW:
                return RED;
            case BLUE:
                return BLUE;
            case FADED:
                return FADED;
            }
            break;
        case FADED:
            return FADED;

        }
        throw new IllegalArgumentException("Unknown color: " + this);
    }
}

File: shared_scala/src/main/scala/se/kth/benchmarks/helpers/ChameneosColour.scala
package se.kth.benchmarks.helpers

sealed trait ChameneosColour {
  def complement(other: ChameneosColour): ChameneosColour;
};
object ChameneosColour {

  def forId(i: Int): ChameneosColour = {
    (i % 3) match {
      case 0 => Red
      case 1 => Yellow
      case 2 => Blue
    }
  }

  case object Red extends ChameneosColour {
    override def complement(other: ChameneosColour): ChameneosColour = other match {
      case Red    => Red
      case Yellow => Blue
      case Blue   => Yellow
      case Faded  => Faded
    }
  }
  case object Yellow extends ChameneosColour {
    override def complement(other: ChameneosColour): ChameneosColour = other match {
      case Red    => Blue
      case Yellow => Yellow
      case Blue   => Red
      case Faded  => Faded
    }
  }
  case object Blue extends ChameneosColour {
    override def complement(other: ChameneosColour): ChameneosColour = other match {
      case Red    => Yellow
      case Yellow => Red
      case Blue   => Blue
      case Faded  => Faded
    }
  }
  case object Faded extends ChameneosColour {
    override def complement(other: ChameneosColour): ChameneosColour = Faded;
  }
}
